### PR TITLE
Unit handling and round-trip .nc conversion fixes

### PR DESCRIPTION
**Pull request by tbody-cfs**

### CODE REVIEW

**Comment by tbody-cfs** [comment]
Closes #10

**Comment by hassec** [comment]
Before merging this I'd like to make sure we have a way to reproduce the error  described in #10. I've tried locally without luck so far and added a test that would catch the in #10 described problem, see #14

**Comment by tbody-cfs** [comment]
Think I've figured out how to reproduce the error. To fix https://github.com/cfs-energy/cfspopcon/actions/runs/7038938605/job/19156931593?pr=7, I added `netcdf4` to the poetry environment. Then I can reproduce #10 with
```
poetry add netcdf4
rm -rf .venv
poetry install
poetry run python tests/regression_results/generate_regression_results.py
poetry run pytest
```

**Comment by tbody-cfs** [comment]
And then I fixed it again https://github.com/cfs-energy/cfspopcon/actions/runs/7039432837/job/19158407447?pr=13 🥳🥳🥳

File: tests/test_unit_handling/test_conversions.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import warnings
```
Comment (hassec) [inline tests/test_unit_handling/test_conversions.py:1]: Thanks for adding this test :)

File: cfspopcon/unit_handling/default_units.py [hunk 2]
```diff
@@ -252,7 +252,7 @@ def convert_to_default_units(value: Quantity, key: str) -> Quantity:
 def convert_to_default_units(value: Union[float, Quantity, xr.DataArray], key: str) -> Union[float, Quantity, xr.DataArray]:
     """Convert an array or scalar to default units."""
     unit = DEFAULT_UNITS[key]
-    if unit is None or unit == "":
+    if unit is None:
```
Comment (hassec) [inline cfspopcon/unit_handling/default_units.py:255]: Nice catch! 👍

File: cfspopcon/file_io.py [hunk 3]
```diff
@@ -47,7 +50,7 @@ def promote_variable(val: xr.DataArray, key: str) -> Any:
     except KeyError:
         pass
 
-    if val.dtype == object:
+    if val.dtype == object or val.dtype.type == np.str_:
```
Comment (hassec) [inline cfspopcon/file_io.py:53]:
```suggestion
    # scipy i/o yields objects for strings while netcdf4 yields np.str_ 
    if val.dtype == object or val.dtype.type == np.str_:
```

**Review by hassec — CHANGES_REQUESTED** [review]
Just a small change request to add a comment to the conversion logic. 

I'd prefer to have updates to the `poetry.lock` in a separate PR but since you already fixed the mypy stuff here let's merge this as is. But in the future we should try and avoid poetry updates in PRs that have other changes.

File: cfspopcon/file_io.py [hunk 4]
```diff
@@ -47,7 +50,8 @@ def promote_variable(val: xr.DataArray, key: str) -> Any:
     except KeyError:
         pass
 
-    if val.dtype == object:
+    # scipy i/o yields objects for strings while netcdf4 yields np.str_
```
Comment (hassec) [inline cfspopcon/file_io.py:53]:
```suggestion
    # scipy i/o yields objects for strings while netcdf4 yields np.str_
```

**Pull request merged**